internal/server/tls.go

```diff
@@ -50,82 +73,86 @@ func tlsConfigFromOptions(
-		return nil, fmt.Errorf("create tls cache: %w", err)
+	if opts.CA == "" || opts.CAPrivateKey == "" {
+		return nil, fmt.Errorf("either a TLS certificate and key or a TLS CA and key is required")
 	}
 
-	manager := &autocert.Manager{
-		Prompt: autocert.AcceptTOS,
-		Cache:  autocert.DirCache(tlsCacheDir),
-	}
-	tlsConfig := manager.TLSConfig()
-	tlsConfig.MinVersion = tls.VersionTLS12
-	// TODO: enabled optional mTLS when opts.CA is set
-	tlsConfig.GetCertificate = SelfSignedOrLetsEncryptCert(manager)
-
-	return tlsConfig, nil
+	cfg.GetCertificate = getCertificate(autocert.DirCache(tlsCacheDir), opts)
+	return cfg, nil
 }
 
-func SelfSignedOrLetsEncryptCert(manager *autocert.Manager) func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
-	return func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
-		ctx := hello.Context()
-		cert, err := manager.GetCertificate(hello)
-		if err == nil {
-			return cert, nil
+func getCertificate(cache autocert.Cache, opts TLSOptions) func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	var lock sync.RWMutex
+
+	getKeyPair := func(ctx context.Context, serverName string) (cert, key []byte) {
+		certBytes, err := cache.Get(ctx, serverName+".crt")
+		if err != nil {
+			logging.S.Warnf("cert: %s", err)
 		}
 
+		keyBytes, err := cache.Get(ctx, serverName+".key")
+		if err != nil {
+			logging.S.Warnf("key: %s", err)
+		}
+		return certBytes, keyBytes
+	}
+
+	return func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		ctx := hello.Context()
 		serverName := hello.ServerName
 
 		if serverName == "" {
 			serverName = hello.Conn.LocalAddr().String()
 		}
 
-		certBytes, err := manager.Cache.Get(ctx, serverName+".crt")
-		if err != nil {
-			logging.S.Warnf("cert: %s", err)
+		lock.RLock()
+		certBytes, keyBytes := getKeyPair(ctx, serverName)
+		lock.RUnlock()
+		if certBytes != nil && keyBytes != nil {
+			return keyPair(certBytes, keyBytes)
 		}
 
-		keyBytes, err := manager.Cache.Get(ctx, serverName+".key")
-		if err != nil {
-			logging.S.Warnf("key: %s", err)
+		lock.Lock()
+		// must check again after write lock is acquired
+		certBytes, keyBytes = getKeyPair(ctx, serverName)
+		defer lock.Unlock()
+		if certBytes != nil && keyBytes != nil {
+			return keyPair(certBytes, keyBytes)
 		}
 
 		// if either cert or key is missing, create it
-		if certBytes == nil || keyBytes == nil {
-			ca, caPrivKey, err := newCA()
-			if err != nil {
-				return nil, err
-			}
+		ca, err := tls.X509KeyPair([]byte(opts.CA), []byte(opts.CAPrivateKey))
+		if err != nil {
+			return nil, err
+		}
 
-			certBytes, keyBytes, err = certs.GenerateCertificate([]string{serverName}, ca, caPrivKey)
-			if err != nil {
-				return nil, err
-			}
+		caCert, err := x509.ParseCertificate(ca.Certificate[0])
+		if err != nil {
+			return nil, err
+		}
 
-			if err := manager.Cache.Put(ctx, serverName+".crt", certBytes); err != nil {
-				return nil, err
-			}
+		hosts := []string{"127.0.0.1", "::1", serverName}
```

Comment:
I change this to always include localhost in the list of valid IP SANs since it's common for a cert to be used on localhost as well.